Several fixes for Xen event channel driver

firscity opened the pull request:
This PR fixes 2 issues with current implementation of Xen events handling:
- use proper barrier to prevent write re-ordering;
- fix cast issues to prevent undefined value during shift with >32 bit value.

Huge thanks @jgrall for help.

firscity committed:
xenvm: evtchn: use proper barrier during events handling

This commit fixes barrier usage in Xen event channel driver. Previously
compiler_barrier() was used and it did not prevent processor from
re-ordering of the write operations, that is needed for correct event
handling. It is now changed to data memory barrier (dmb()), which will
work as expected in this situation.

Thanks @jgrall for suggestion.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

firscity committed:
xenvm: evtchn: fix undefined value during event handling

This commit fixes issue with undefined value during position calculations.
Shift for positions may be >32, so we need to explicitly use unsigned long
casting to prevent unexpected behavior during event handling.

Thanks @jgrall for suggestion.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

carlocaione commented on drivers/xen/events.c:
```diff
@@ -218,7 +218,7 @@ static void events_isr(void *data)
 	vcpu->evtchn_upcall_pending = 0;
 
-	compiler_barrier();
+	__DMB();
```
We are trying to move away from `__XXX`. Can you use `dmb()` instead?

firscity replied:
Sure, thanks for review! Fixed.

carlocaione commented on drivers/xen/events.c:
```diff
@@ -227,12 +227,12 @@ static void events_isr(void *data)
 		/* Find first position, clear it in selector and process */
 		pos_index = __builtin_ffsl(pos_selector) - 1;
-		pos_selector &= ~(1 << pos_index);
+		pos_selector &= ~(((xen_ulong_t) 1) << pos_index);
```
out of curiosity, why not using `1ULL` o `1UL` nstead?

firscity replied:
Such version will be more consistent with Xen approach. AFAIK definition of `xen_ulong_t` and structure, where events are stored various on different archs. So, `1UUL` will be OK for arm64, but probably will be a problem if someone will try to use it with x86 in future.

jgrall replied:
> Such version will be more consistent with Xen approach. AFAIK definition of xen_ulong_t and structure, where events are stored various on different archs. So, 1UUL will be OK for arm64, but probably will be a problem if someone will try to use it with x86 in future.

So on x86, ``xen_ulong_t`` is effectively an ``unsigned long``. But this as some consequences because the 32-bit and 64-bit structure layout will be different and this requires a compat layer to support 32-bit guest on 64-bit hypervisor.

On Arm, we decided to go with ``uint64_t`` so the ABI doesn't change between Arm32 an Arm64.

At the moment, we technically use``1ULL``. However, I think is better to cast to the same type as the resulting type. This would help if someone decide to define ``xen_ulong_t`` as something bigger than ``unsigned long long``.

luca-fancellu reviewed (dismissed):
Many thanks for that!

jgrall commented:
> Previously compiler_barrier() was used and it did not prevent re-ordering of write
operations, that is needed for correct event handling.

NIT: ``compiler_barrier()`` prevents re-ordering at the compiler level. However, it doesn't prevent re-ordering by the processor. So I would suggest to add the word ``processor``.

firscity commented:
> NIT: compiler_barrier() prevents re-ordering at the compiler level. However, it doesn't prevent re-ordering by the processor. So I would suggest to add the word processor.

Agree. I meant that, but wrongly explain this. Fixed.
Also, changed `__DMB()` to `dmb()` in commit message during rewording to be in sync with changes.

luca-fancellu commented:
@firscity I think you need to wrap your commit message to max 75 chars:

> Commit aa12b4f36c:
> 4: UC4 Line exceeds max length (77>75): "compiler_barrier() was used and it did not prevent processor from re-ordering"
> 5: UC4 Line exceeds max length (77>75): "of the write operations, that is needed for correct event handling. It is now"

firscity commented:
@luca-fancellu, oh, sorry, fixed.

luca-fancellu reviewed (approved):
Now that the CI is happy, I think this can find its way. @povergoing
